File: crates/printobserver/tests/support/interposer.rs
//! What one invocation's own process tree did, recorded from inside it.
//!
//! # Why this exists beside `strace`
//!
//! On Linux this tier observes a program with `strace`. macOS has no `strace`,
//! and its own tracer, `dtruss`, needs System Integrity Protection turned off,
//! which no hosted runner has. So there the observation is made from inside the
//! program instead: `interposer.c`, built here with the system's own `cc` into
//! the run's scratch directory, is loaded into the program through
//! `DYLD_INSERT_LIBRARIES` and replaces `connect` and the file-access calls
//! through dyld's interposing table. The variable is inherited, so every child
//! the program starts is observed too, exactly as `strace -f` follows one.
//!
//! # What SIP strips, and why that does not weaken the claim
//!
//! dyld ignores every `DYLD_*` variable for a binary SIP protects — anything
//! under `/usr/bin` or `/System`, say — and for a program signed with the
//! hardened runtime. A process of that kind that the program started would run
//! unobserved. What this tier holds is a property of **this program** and of the
//! processes it starts itself, and the program under test is a cargo build that
//! is not hardened-runtime signed; the load is not assumed either, because the
//! library records `loaded <pid>` from its constructor and a run whose own
//! process did not record one panics rather than reading as a run that did
//! nothing. The client commands start no system binary to reach anything: the
//! one place a connection is opened is this program's own transport.
//!
//! # One library, both platforms
//!
//! The same source builds on Linux, loaded with `LD_PRELOAD`, so the recording
//! and the reading of it are proven on every host this tier runs on — not only
//! on the runners where they are the only observation there is.

use std::path::{Path, PathBuf};
use std::process::{Command, Output, Stdio};
use std::sync::atomic::{AtomicU64, Ordering};

/// The interposer's source, compiled into each scratch directory that needs it.
const SOURCE: &str = include_str!("interposer.c");

/// The variable naming the file the interposer appends what it saw to.
const LOG_ENV: &str = "PRINTOBSERVER_INTERPOSE_LOG";

/// The variable naming the paths it refuses, one per line.
const WATCH_ENV: &str = "PRINTOBSERVER_INTERPOSE_WATCH";

/// The variable the dynamic loader reads a library to insert from.
#[cfg(target_os = "macos")]
const LOADER_ENV: &str = "DYLD_INSERT_LIBRARIES";
/// The variable the dynamic loader reads a library to insert from.
#[cfg(not(target_os = "macos"))]
const LOADER_ENV: &str = "LD_PRELOAD";

/// The programs the loader will not insert a library into.
#[cfg(target_os = "macos")]
const IGNORED_FOR: &str = "a SIP-protected program and a hardened-runtime one";
/// The programs the loader will not insert a library into.
#[cfg(not(target_os = "macos"))]
const IGNORED_FOR: &str = "a set-user-ID program and a statically linked one";

/// The file the built library is kept under.
#[cfg(target_os = "macos")]
const LIBRARY: &str = "interposer.dylib";
/// The file the built library is kept under.
#[cfg(not(target_os = "macos"))]
const LIBRARY: &str = "interposer.so";

/// What `cc` is told before the source: build a library to be inserted.
#[cfg(target_os = "macos")]
const LINKED_AS: &[&str] = &["-dynamiclib"];
/// What `cc` is told before the source: build a library to be inserted.
#[cfg(not(target_os = "macos"))]
const LINKED_AS: &[&str] = &["-shared", "-fPIC"];

/// What `cc` is told after the source: the libraries it calls into.
#[cfg(target_os = "macos")]
const LINKED_WITH: &[&str] = &[];
/// What `cc` is told after the source: the libraries it calls into.
#[cfg(not(target_os = "macos"))]
const LINKED_WITH: &[&str] = &["-ldl"];

/// How one run's own files are named apart from every other's.
static RUNS: AtomicU64 = AtomicU64::new(0);

/// One run under the interposer.
pub struct Interposed {
    /// What the program exited with and printed.
    pub output: Output,
    /// Every line the interposer recorded for the process tree, with the lines
    /// saying where it was loaded taken out: empty when the tree connected to
    /// nothing and touched no watched path.
    pub activity: String,
}

impl Interposed {
    /// Every access to a watched path the interposer refused, one per line and
    /// with every connection taken out: empty when the tree touched no watched
    /// path, and naming the path and `EACCES` for each one it did.
    pub fn touched(&self) -> String {
        let mut touched = String::new();
        for line in self.activity.lines() {
            if !line.starts_with("connected ") {
                touched.push_str(line);
                touched.push('\n');
            }
        }
        touched
    }
}

/// Run one command with the interposer loaded into it, refusing every access
/// to a watched path.
///
/// # Panics
///
/// Panics when the interposer cannot be built, when the command cannot be
/// started, or when the interposer was not loaded into the command's own
/// process — each of which is a run whose recording says nothing about it.
pub fn interposed(mut command: Command, scratch: &Path, watching: &[PathBuf]) -> Interposed {
    let library = built(scratch);
    let log = scratch.join(unique("interposed", "log"));
    std::fs::write(&log, "").expect("the interposer's log is writable");
    command
        .env(LOADER_ENV, &library)
        .env(LOG_ENV, &log)
        .env(WATCH_ENV, watched(watching))
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    let child = command
        .spawn()
        .unwrap_or_else(|error| panic!("the program could not be started to be observed: {error}"));
    let root = child.id();
    let output = child
        .wait_with_output()
        .unwrap_or_else(|error| panic!("the observed program could not be waited for: {error}"));
    let recorded = std::fs::read_to_string(&log)
        .unwrap_or_else(|error| panic!("the interposer's log {} reads: {error}", log.display()));
    let activity = activity(&recorded, root).unwrap_or_else(|| {
        panic!(
            "the interposer at {} was not loaded into process {root}, so nothing it recorded \
             says what that process did. {LOADER_ENV} is ignored for {IGNORED_FOR}; this \
             tier's claims about what the program reaches rest on the observation. It \
             said:\n{}{}",
            library.display(),
            String::from_utf8_lossy(&output.stdout),
            String::from_utf8_lossy(&output.stderr),
        )
    });
    Interposed { output, activity }
}

/// What a recording says the tree did, when the recording is of `root`.
///
/// `None` when the interposer never said it was loaded into that process.
pub fn activity(recorded: &str, root: u32) -> Option<String> {
    let root = root.to_string();
    let mut loaded = false;
    let mut kept = String::new();
    for line in recorded.lines() {
        if let Some(process) = line.strip_prefix("loaded ") {
            loaded |= process == root;
        } else {
            kept.push_str(line);
            kept.push('\n');
        }
    }
    loaded.then_some(kept)
}

/// Every watched path as the interposer reads them: each as given and, where it
/// differs, as the filesystem resolves it, one per line.
fn watched(watching: &[PathBuf]) -> String {
    let mut named: Vec<String> = Vec::new();
    for path in watching {
        named.push(path.display().to_string());
        if let Ok(resolved) = std::fs::canonicalize(path)
            && resolved != *path
        {
            named.push(resolved.display().to_string());
        }
    }
    named.join("\n")
}

/// The interposer, built into one scratch directory once.
fn built(scratch: &Path) -> PathBuf {
    let library = scratch.join(LIBRARY);
    if library.is_file() {
        return library;
    }
    let source = scratch.join(unique("interposer", "c"));
    std::fs::write(&source, SOURCE).expect("the interposer's source is writable");
    let staged = scratch.join(unique("interposer", "staged"));
    let compiled = Command::new("cc")
        .args(LINKED_AS)
        .arg("-o")
        .arg(&staged)
        .arg(&source)
        .args(LINKED_WITH)
        .output()
        .unwrap_or_else(|error| {
            panic!(
                "`cc` could not run to build the interposer, and this tier's observation of \
                 what a run reaches rests on it: {error}"
            )
        });
    assert!(
        compiled.status.success(),
        "`cc` did not build the interposer, and this tier's observation of what a run reaches \
         rests on it:\n{}",
        String::from_utf8_lossy(&compiled.stderr)
    );
    // Renamed into place whole, so a run beside this one never loads half a file.
    std::fs::rename(&staged, &library).expect("the built interposer moves into place");
    library
}

/// A file name no other run in this process or any other shares.
fn unique(stem: &str, extension: &str) -> String {
    format!(
        "{stem}-{}-{}.{extension}",
        std::process::id(),
        RUNS.fetch_add(1, Ordering::Relaxed)
    )
}

File: crates/printobserver/tests/support/interposer.c
/*
 * What one invocation's own process tree did, recorded from inside it.
 *
 * This is the command-line tier's observation where `strace` is not there to
 * make it: loaded into the program under test (and, inherited through the
 * environment, into every child it starts), it records every `connect` any of
 * them makes and, given paths to watch, every file access to one of those
 * paths, failing each with EACCES as a user who may not read them would be
 * failed. `interposer.rs` builds it, loads it and reads what it wrote.
 *
 * `PRINTOBSERVER_INTERPOSE_SELF`, when set, confines it to the process it was
 * loaded into: the loader variable is taken out of that process's environment
 * before it starts anything, so its children carry no recorder.
 *
 * Three kinds of line are written, one `write` each, appended to the file
 * `PRINTOBSERVER_INTERPOSE_LOG` names:
 *
 *   loaded <pid>                       this library is in process <pid>
 *   connected <address>:<port>         a connect to an internet endpoint
 *   <call> "<path>" = -1 EACCES        an access to a watched path, refused
 *
 * `PRINTOBSERVER_INTERPOSE_WATCH` names the watched paths, one per line; an
 * access to one of them or to anything beneath one is refused.
 *
 * On macOS the wrappers are installed through dyld's `__DATA,__interpose`
 * table, which rebinds every other image's references to the named functions
 * and leaves this library's own calls reaching the originals. On Linux the
 * same wrappers are exported under the functions' own names and loaded with
 * `LD_PRELOAD`, calling through to the next definition; that build is what
 * lets the recording half be proven on a Linux host as well.
 *
 * The functions wrapped are the ones Rust's standard library and the `libc`
 * crate reach for these operations on each platform: `open`, `openat`, `stat`,
 * `lstat`, `fstatat`, `access`, `faccessat`, `opendir` and `connect`, plus on
 * Linux the large-file and `statx` spellings glibc links them under. The
 * table below names each by its C spelling, so it interposes whatever symbol
 * the SDK's own headers resolve that name to — on Apple silicon, the one
 * supported macOS, that is the plain symbol.
 */

#ifndef __APPLE__
#define _GNU_SOURCE
#endif

#include <arpa/inet.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <netinet/in.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#ifdef __APPLE__
#include <sys/param.h>
/* The original, reached directly: dyld does not rebind this image's own calls. */
#define REAL(name) name
/* A wrapper is this library's own, and reached only through the table. */
#define WRAPPER static
#define WRAP(name) po_##name
#define INTERPOSE(name)                                                        \
    __attribute__((used)) static struct {                                      \
        const void *replacement;                                               \
        const void *replacee;                                                  \
    } po_interpose_##name __attribute__((section("__DATA,__interpose"))) = {   \
        (const void *)(unsigned long)&po_##name,                               \
        (const void *)(unsigned long)&name,                                    \
    };
#else
#include <dlfcn.h>
/* The next definition after this library's own, which is the C library's. */
static void *po_next(const char *name) {
    void *found = dlsym(RTLD_NEXT, name);
    if (found == NULL) {
        abort();
    }
    return found;
}
#define REAL(name) ((__typeof__(&name))po_next(#name))
#define WRAPPER __attribute__((visibility("default")))
#define WRAP(name) name
#endif

#define PO_LOG_ENV "PRINTOBSERVER_INTERPOSE_LOG"
#define PO_WATCH_ENV "PRINTOBSERVER_INTERPOSE_WATCH"
/* Set, this process alone is observed: see po_loaded. */
#define PO_SELF_ENV "PRINTOBSERVER_INTERPOSE_SELF"
#ifdef __APPLE__
#define PO_LOADER_ENV "DYLD_INSERT_LIBRARIES"
#else
#define PO_LOADER_ENV "LD_PRELOAD"
#endif
#define PO_PATH_CAP 4096

/* Append one line to the log, leaving errno as the caller had it. */
static void po_log(const char *line, size_t length) {
    int saved = errno;
    const char *log = getenv(PO_LOG_ENV);
    if (log != NULL && *log != '\0') {
        int fd = REAL(open)(log, O_WRONLY | O_APPEND | O_CREAT | O_CLOEXEC, 0600);
        if (fd >= 0) {
            ssize_t written = write(fd, line, length);
            (void)written;
            close(fd);
        }
    }
    errno = saved;
}

/*
 * Asked to observe this process alone, the library takes its own loader
 * variable out of the environment here, before the program has started
 * anything, so nothing it starts inherits the library. A journey whose subject
 * is one program's own calls asks for this: a child of another ABI — the
 * system shell on Apple Silicon is arm64e where this library is arm64 — would
 * otherwise be killed by the loader for a library it cannot take.
 */
__attribute__((constructor)) static void po_loaded(void) {
    char line[64];
    if (getenv(PO_SELF_ENV) != NULL) {
        unsetenv(PO_LOADER_ENV);
    }
    int length = snprintf(line, sizeof line, "loaded %ld\n", (long)getpid());
    if (length > 0 && (size_t)length < sizeof line) {
        po_log(line, (size_t)length);
    }
}

/* The absolute path one call names, relative paths resolved against their directory. */
static int po_resolve(int dirfd, const char *path, char *out, size_t cap) {
    if (path == NULL || *path == '\0') {
        return 0;
    }
    if (*path == '/') {
        int length = snprintf(out, cap, "%s", path);
        return length > 0 && (size_t)length < cap;
    }
    char directory[PO_PATH_CAP];
    if (dirfd == AT_FDCWD) {
        if (getcwd(directory, sizeof directory) == NULL) {
            return 0;
        }
    } else {
#ifdef __APPLE__
        char named[MAXPATHLEN];
        if (fcntl(dirfd, F_GETPATH, named) == -1) {
            return 0;
        }
        snprintf(directory, sizeof directory, "%s", named);
#else
        char link[64];
        snprintf(link, sizeof link, "/proc/self/fd/%d", dirfd);
        ssize_t length = readlink(link, directory, sizeof directory - 1);
        if (length <= 0) {
            return 0;
        }
        directory[length] = '\0';
#endif
    }
    int length = snprintf(out, cap, "%s/%s", directory, path);
    return length > 0 && (size_t)length < cap;
}

/* Whether a path is one of the watched ones, or beneath one. */
static int po_watched(const char *path) {
    const char *watching = getenv(PO_WATCH_ENV);
    if (watching == NULL) {
        return 0;
    }
    const char *entry = watching;
    while (*entry != '\0') {
        const char *end = strchr(entry, '\n');
        size_t length = end == NULL ? strlen(entry) : (size_t)(end - entry);
        while (length > 1 && entry[length - 1] == '/') {
            length--;
        }
        if (length > 0 && strncmp(path, entry, length) == 0 &&
            (path[length] == '\0' || path[length] == '/')) {
            return 1;
        }
        if (end == NULL) {
            break;
        }
        entry = end + 1;
    }
    return 0;
}

/* Refuse one call on a watched path, recording it; answer whether it was refused. */
static int po_refused(const char *call, int dirfd, const char *path) {
    char resolved[PO_PATH_CAP];
    if (!po_resolve(dirfd, path, resolved, sizeof resolved) || !po_watched(resolved)) {
        return 0;
    }
    char line[PO_PATH_CAP + 64];
    int length = snprintf(line, sizeof line, "%s \"%s\" = -1 EACCES\n", call, resolved);
    if (length > 0 && (size_t)length < sizeof line) {
        po_log(line, (size_t)length);
    }
    errno = EACCES;
    return 1;
}

/* Record the internet endpoint one connect names. */
static void po_connected(const struct sockaddr *address) {
    char host[INET6_ADDRSTRLEN];
    char line[INET6_ADDRSTRLEN + 32];
    int length = 0;
    if (address == NULL) {
        return;
    }
    if (address->sa_family == AF_INET) {
        const struct sockaddr_in *v4 = (const struct sockaddr_in *)address;
        if (inet_ntop(AF_INET, &v4->sin_addr, host, sizeof host) == NULL) {
            return;
        }
        length = snprintf(line, sizeof line, "connected %s:%u\n", host,
                          (unsigned)ntohs(v4->sin_port));
    } else if (address->sa_family == AF_INET6) {
        const struct sockaddr_in6 *v6 = (const struct sockaddr_in6 *)address;
        if (inet_ntop(AF_INET6, &v6->sin6_addr, host, sizeof host) == NULL) {
            return;
        }
        length = snprintf(line, sizeof line, "connected [%s]:%u\n", host,
                          (unsigned)ntohs(v6->sin6_port));
    }
    if (length > 0 && (size_t)length < sizeof line) {
        po_log(line, (size_t)length);
    }
}

/* Whether an open carries a mode argument: only one that may create a file does. */
static int po_creates(int flags) {
#ifdef O_TMPFILE
    if ((flags & O_TMPFILE) == O_TMPFILE) {
        return 1;
    }
#endif
    return (flags & O_CREAT) != 0;
}

WRAPPER int WRAP(connect)(int fd, const struct sockaddr *address, socklen_t length) {
    po_connected(address);
    return REAL(connect)(fd, address, length);
}

WRAPPER int WRAP(open)(const char *path, int flags, ...) {
    mode_t mode = 0;
    if (po_creates(flags)) {
        va_list arguments;
        va_start(arguments, flags);
        mode = (mode_t)va_arg(arguments, int);
        va_end(arguments);
    }
    if (po_refused("open", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(open)(path, flags, mode);
}

WRAPPER int WRAP(openat)(int dirfd, const char *path, int flags, ...) {
    mode_t mode = 0;
    if (po_creates(flags)) {
        va_list arguments;
        va_start(arguments, flags);
        mode = (mode_t)va_arg(arguments, int);
        va_end(arguments);
    }
    if (po_refused("openat", dirfd, path)) {
        return -1;
    }
    return REAL(openat)(dirfd, path, flags, mode);
}

WRAPPER int WRAP(stat)(const char *path, struct stat *buffer) {
    if (po_refused("stat", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(stat)(path, buffer);
}

WRAPPER int WRAP(lstat)(const char *path, struct stat *buffer) {
    if (po_refused("lstat", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(lstat)(path, buffer);
}

WRAPPER int WRAP(fstatat)(int dirfd, const char *path, struct stat *buffer, int flags) {
    if (po_refused("fstatat", dirfd, path)) {
        return -1;
    }
    return REAL(fstatat)(dirfd, path, buffer, flags);
}

WRAPPER int WRAP(access)(const char *path, int mode) {
    if (po_refused("access", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(access)(path, mode);
}

WRAPPER int WRAP(faccessat)(int dirfd, const char *path, int mode, int flags) {
    if (po_refused("faccessat", dirfd, path)) {
        return -1;
    }
    return REAL(faccessat)(dirfd, path, mode, flags);
}

WRAPPER DIR *WRAP(opendir)(const char *path) {
    if (po_refused("opendir", AT_FDCWD, path)) {
        return NULL;
    }
    return REAL(opendir)(path);
}

#ifdef __APPLE__
INTERPOSE(connect)
INTERPOSE(open)
INTERPOSE(openat)
INTERPOSE(stat)
INTERPOSE(lstat)
INTERPOSE(fstatat)
INTERPOSE(access)
INTERPOSE(faccessat)
INTERPOSE(opendir)
#else
WRAPPER int WRAP(open64)(const char *path, int flags, ...) {
    mode_t mode = 0;
    if (po_creates(flags)) {
        va_list arguments;
        va_start(arguments, flags);
        mode = (mode_t)va_arg(arguments, int);
        va_end(arguments);
    }
    if (po_refused("open64", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(open64)(path, flags, mode);
}

WRAPPER int WRAP(openat64)(int dirfd, const char *path, int flags, ...) {
    mode_t mode = 0;
    if (po_creates(flags)) {
        va_list arguments;
        va_start(arguments, flags);
        mode = (mode_t)va_arg(arguments, int);
        va_end(arguments);
    }
    if (po_refused("openat64", dirfd, path)) {
        return -1;
    }
    return REAL(openat64)(dirfd, path, flags, mode);
}

WRAPPER int WRAP(stat64)(const char *path, struct stat64 *buffer) {
    if (po_refused("stat64", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(stat64)(path, buffer);
}

WRAPPER int WRAP(lstat64)(const char *path, struct stat64 *buffer) {
    if (po_refused("lstat64", AT_FDCWD, path)) {
        return -1;
    }
    return REAL(lstat64)(path, buffer);
}

WRAPPER int WRAP(fstatat64)(int dirfd, const char *path, struct stat64 *buffer, int flags) {
    if (po_refused("fstatat64", dirfd, path)) {
        return -1;
    }
    return REAL(fstatat64)(dirfd, path, buffer, flags);
}

WRAPPER int WRAP(statx)(int dirfd, const char *path, int flags, unsigned int mask,
                   struct statx *buffer) {
    if (po_refused("statx", dirfd, path)) {
        return -1;
    }
    return REAL(statx)(dirfd, path, flags, mask, buffer);
}
#endif
